language/move-prover/bytecode-to-boogie/src/spec_translator.rs

```diff
@@ -284,30 +467,224 @@ impl<'a> SpecTranslator<'a> {
+        let arg_index = {
+            self.function_info
+                .arg_names
+                .iter()
+                .position(|s| s == name)
+                .unwrap_or_else(|| {
+                    self.error(
+                        &format!("cannot determine position of parameter `{}`", name),
+                        0,
+                    )
+                })
+        };
+        // Get the parameter type.
+        let arg_sig = {
+            if let Some(ty) = self.get_param_types().get(arg_index) {
+                ty.clone()
+            } else {
                 self.error(
-                    &format!("cannot determine position of parameter `{}`", name),
-                    0,
+                    &format!("cannot determine type of parameter {}", name),
+                    ERROR_TYPE,
                 )
-            })
+            }
+        };
+        BoogieExpr(format!("arg{}", arg_index), arg_sig)
+    }
+
+    /// Translate a function return value.
+    fn translate_return(&mut self) -> BoogieExpr {
+        let sig = self.get_function_signature();
+        if sig.return_count() > 1 {
+            // TODO: the parser currently only handles `return` for single value. Need to
+            //   generalize for multiple ret0, ret1, ...
+            self.error(
+                "cannt handle more than one return value as of now",
+                BoogieExpr("<ret>".to_string(), SignatureToken::U64),
+            )
+        } else if sig.return_count() < 1 {
+            self.error(
+                "function does not return a value",
+                BoogieExpr("<ret>".to_string(), SignatureToken::U64),
+            )
+        } else {
+            BoogieExpr(
+                "ret0".to_string(),
+                sig.return_tokens()
+                    .nth(0)
+                    .expect("non-empty return")
+                    .signature_token()
+                    .clone(),
+            )
+        }
+    }
+
+    /// Translates a field name, where `sig` is the type from which the field is selected.
+    /// Returns boogie field name and type.
+    fn translate_field(
+        &mut self,
+        mut sig: &SignatureToken,
+        field: &Field,
+    ) -> (String, SignatureToken) {
+        // If this is a reference, use the underlying type. This function works with both
+        // references and non-references.
+        let is_ref = if let SignatureToken::Reference(s) = sig {
+            sig = &*s;
+            true
+        } else {
+            false
+        };
+        if let SignatureToken::Struct(handle_idx, type_args) = sig {
+            let struct_handle = self.module.struct_handle_at(*handle_idx);
+            // We cannot lookup type information for fields of imported modules via the handle view.
+            // Instead we need to retrieve the struct definition, which might be in another module.
+            // TODO: validate that I've not missed some feature in file_format.rs. It might be
```

Comment:
This is an interesting problem. It's related to the one about needing a richer grammar for (e.g.) `exists<>` in the spec language than in the bytecode language.

- `BorrowField<f>` can only be performed on fields of resources declared in the current module. Thus, the field definition table of the module only talks about its own fields, not about the fields of its dependencies.
- However: you clearly want postconditions to talk about fields of different modules.  e.g.,

```
module M1 {
  resource R { b: bool }
  public flip(r: &mut R) {
    *r.b = !r.b
  }
}

module M2 {
  resource Wrapper {  r: M.R }
  public call_ext(w: &mut Self.Wrapper) {
    M1.flip(w)
  } // post: w/r/b = !old(w/r/b)
}
```

This is inconvenient because there's no representation of the field `b` anywhere in the bytecode for `M2`. We'll need to look through all of the bytecode for the dependencies of `M2` in order to fully typecheck the specs...

Reply:
Yep, that is what I need to do here. But because we do not have separate compilation anyway, it wasn't too much of a problem.